### PR TITLE
Workshop1 POC

### PR DESCRIPTION
**Pull request by alexander-smityuk**
Workshop #1 for UI Architecture

### CODE REVIEW

File: app/src/main/java/com/android/fundamentals/workshop01/Workshop1Fragment.kt [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        initViews(view)
+        setUpListeners()
+
+        // Attaching view to the presenter
+        presenter.attachView(this)
+    }
+
+    override fun onDestroyView() {
+        // Detaching view from the presenter
+        presenter.detachView()
```
Comment (rHlushan) [inline app/src/main/java/com/android/fundamentals/workshop01/Workshop1Fragment.kt:37]: comment it also to put attention?

File: app/src/main/java/com/android/fundamentals/workshop01/Workshop1Fragment.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    private var loginSuccess: View? = null
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        initViews(view)
+        setUpListeners()
+
+        // Attaching view to the presenter
+        presenter.attachView(this)
```
Comment (rHlushan) [inline app/src/main/java/com/android/fundamentals/workshop01/Workshop1Fragment.kt:32]: comment it also to put attention?

File: app/src/main/java/com/android/fundamentals/workshop01/Workshop1Presenter.kt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    private val mainDispatcher: CoroutineDispatcher
+) {
+
+    private val presenterScope: CoroutineScope =
+        CoroutineScope(SupervisorJob() + mainDispatcher)
+
+    private var view: Workshop1View? = null
+
+    fun attachView(view: Workshop1View) {
+        this.view = view
```
Comment (rHlushan) [inline app/src/main/java/com/android/fundamentals/workshop01/Workshop1Presenter.kt:17]: comment it also to put attention?

File: app/src/main/java/com/android/fundamentals/workshop01/Workshop1Presenter.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    fun detachView() {
+        presenterScope.cancel()
+        this.view = null
+    }
```
Comment (rHlushan) [inline app/src/main/java/com/android/fundamentals/workshop01/Workshop1Presenter.kt:23]: comment it also to put attention?

File: app/src/main/java/com/android/fundamentals/workshop01/Workshop1View.kt [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.android.fundamentals.workshop01
+
+interface Workshop1View {
```
Comment (rHlushan) [inline app/src/main/java/com/android/fundamentals/workshop01/Workshop1View.kt:3]: remove Workshop1SolutionView and use this one?

File: app/src/test/java/com/android/foundamentals/workshop01/Workshop1ViewMock.kt [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package com.android.foundamentals.workshop01
+
+import com.android.fundamentals.workshop01.Workshop1View
+
+internal class Workshop1ViewMock : Workshop1View {
```
Comment (rHlushan) [inline app/src/test/java/com/android/foundamentals/workshop01/Workshop1ViewMock.kt:5]: remove Workshop1SolutionViewMock and use this one?

File: app/src/main/java/com/android/fundamentals/workshop01/Workshop1Fragment.kt [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    override fun setLoading(loading: Boolean) {
+        //TODO 01: Make loader visible/gone and loginBtn enable/disable
+    }
+
+    override fun showUserNameError() {
+        //TODO 02: Set error for userNameInput from strings resources
+    }
+
+    override fun showPasswordError() {
+        //TODO 03: Set error for passwordInput from strings resources
```
Comment (rHlushan) [inline app/src/main/java/com/android/fundamentals/workshop01/Workshop1Fragment.kt]: add resource witch should be used?

File: app/src/main/java/com/android/fundamentals/workshop01/Workshop1Fragment.kt [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+        super.onDestroyView()
+    }
+
+    override fun setLoading(loading: Boolean) {
+        //TODO 01: Make loader visible/gone and loginBtn enable/disable
+    }
+
+    override fun showUserNameError() {
+        //TODO 02: Set error for userNameInput from strings resources
```
Comment (rHlushan) [inline app/src/main/java/com/android/fundamentals/workshop01/Workshop1Fragment.kt]: add resource witch should be used?

File: app/src/test/java/com/android/foundamentals/workshop01/Workshop1PresenterTest.kt [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        )
+
+        // Call presenter method to attach view
+        // TODO 02: Uncomment this code and run the test
+        //presenter.login(userName = "", password = "")
+
+        // Checking the view
+        view.assert(
+            expectedLoading = true,
+            expectedUserNameErrorShownTimes = 0,
```
Comment (rHlushan) [inline app/src/test/java/com/android/foundamentals/workshop01/Workshop1PresenterTest.kt:52]: expectedUserNameErrorShownTimes = TODO(),
             expectedPasswordErrorShownTimes = TODO(),
             expectedSuccessShownTimes = TODO(),
with comments?
and same for all cases below?

File: app/src/main/java/com/android/fundamentals/workshop01/Workshop1Fragment.kt [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    override fun showPasswordError() {
+        //TODO 03: Set error for passwordInput from strings resources
+    }
+
+    override fun showSuccess() {
+        //TODO 04: Hide logicBtn and show loginSuccess
```
Comment (vadzim-vys) [inline app/src/main/java/com/android/fundamentals/workshop01/Workshop1Fragment.kt]: logicBtn -> loginBtn

File: app/src/main/java/com/android/fundamentals/workshop01/Workshop1Fragment.kt [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        loginBtn?.setOnClickListener {
+            tryToLogin()
+        }
+    }
+
+    private fun tryToLogin() {
+        val inputUserName = userNameInput?.text?.toString().orEmpty()
+        val inputPassword = passwordInput?.text?.toString().orEmpty()
+
+        //TODO 04: Call presenter login method
```
Comment (vadzim-vys) [inline app/src/main/java/com/android/fundamentals/workshop01/Workshop1Fragment.kt]: TODO 04 -> TODO 05

File: app/src/main/java/com/android/fundamentals/workshop01/Workshop1Presenter.kt [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+    fun login(userName: String, password: String) {
+        presenterScope.launch {
+            //TODO 01: Set loading to true
```
Comment (vadzim-vys) [inline app/src/main/java/com/android/fundamentals/workshop01/Workshop1Presenter.kt]: should be TODO 6

**Pull request closed without merging**